lib/datadog/appsec/utils/read_write_lock.rb

```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+module Datadog
+  module AppSec
+    module Utils
+      # Simple implentation for a fair read write lock
+      # It allows for as many concurrent readers but only one concurrent writer.
+      # To avoid the writer thread being exhausted, once a writer thread has signaled that it wants to acquire the lock,
+      # no more reader threads can acquire the lock.
+      # The new reader threads will wait for the writer thread to finish.
+      class ReadWriteLock
+        def initialize
+          @reader_mutex    = Mutex.new
+          @reader_q        = ConditionVariable.new
+          @reader_count    = 0
+          @reader_releases = 0
+
+          @writer_mutex    = Mutex.new
+          @writer_q        = ConditionVariable.new
+          @writter         = false
+        end
+
+        def with_rlock
+          rlock
+          yield
+        ensure
+          runlock
+        end
+
+        def with_lock
+          lock
+          yield
+        ensure
+          unlock
+        end
```

Comment:
This API is inspired by:

- the [Go API's RWMutex](https://pkg.go.dev/sync#RWMutex)
- the Ruby API's `Mutex`

Which seemed like very close relatives.

Both `concurrent_ruby` and the Java API veer off significantly from Ruby's `Mutex` interface.

Regarding that one:

```
lock.read_lock.synchronize { puts "Hey I'm reading!" }
lock.write_lock.synchronize { puts "Hey, I'm writing!" }
```

This method chaining feels very Rails-y and would need two method receiving intermediate facade instances for the sole purpose of delegating to the parent one, especially since internal state is by nature shared. Probably if you want to keep a single `synchronize` instead of `with_*` then `lock.synchronize(:read) { }` would suffice, with the default argument value being `:write`, thus making that `RWMutex` a drop-in replacement for `Mutex`.

Reply:
I like the `lock.synchronize(:read) { }` API